swift4: Fixes for closure arguments

spevans opened the pull request:
- Fix destructuring

- Use correct number of arguments in closures

Compatible with swift 3 & 4

pushkarnk commented:
@swift-ci test

parkera commented:
I'm not sure I understand what was wrong with the previous syntax. Why did this become a problem?

spevans commented:
There were 2 sets of errors:

1. The closure argument count needs to match the signature, eg ` Contextual closure type '(UnsafeMutableRawPointer, Int) -> Void' expects 2 arguments, but 1 was used in closure body` and `Contextual type for closure argument list expects 1 argument, which cannot be implicitly ignored`

2. Some cases for { (key, value)  needed to be replaced with { (arg) ; let (key, value) = arg to solve the following: 
 `Closure tuple parameter '(Notification, [RunLoopMode])' does not support destructuring`

Most of the fixes were from fix-its in Xcode

damuellen commented:
Just yesterday, the error message "..does not support destructuring" occurred to me in some places in one of my projects. The Xcode fix-it has helped, but I would also like to know why Swift 4 is different in this case than Swift 3.

spevans commented:
Its due to SE-110 https://github.com/apple/swift-evolution/blob/master/proposals/0110-distingish-single-tuple-arg.md 
I think it made an incorrect argument count an  error in Swift4

parkera commented:
Is there an alternative to effectively re-declaring the argument all over the place? This can't be the pattern we really want people to use.

parkera commented:
I sent a message to the swift evolution mailing list about this. I think we should have a further discussion before we accept this; it seems like a dramatic reduction in usability of the language in general.

spevans commented:
Yeah thats a good idea. It does seem quite a source breaking change given how much closures are used in general

xwu commented on Foundation/Dictionary.swift:
```diff
@@ -18,7 +18,8 @@ extension Dictionary : _ObjectTypeBridgeable {
         var idx = 0
         
-        self.forEach { (keyItem, valueItem) in
+        self.forEach { (arg) in
```
Stylistic point: this and all subsequent parameter lists in closures don't need the parentheses around `arg`.

spevans replied:
This change actually came from the migrator, it may be worth fixing that to not add parentheses as I was taking the fix-its as best practice. Although I'll see what the outcome of the discussion on swift-evolution is before making any changes to this PR

guidomb replied:
Sorry if this is not the best place to ask this question but what about overloading functions like `forEach` or `filter` to do something like https://gist.github.com/guidomb/b313667616612fc3e0c74c02a066dd8b and avoid having to unpack tuples

xwu commented on Foundation/NSIndexSet.swift:
```diff
@@ -92,8 +92,8 @@ open class NSIndexSet : NSObject, NSCopying, NSMutableCopying, NSSecureCoding {
         let set = NSMutableIndexSet()
-        enumerateRanges(options: []) {
-            set.add(in: $0.0)
+        enumerateRanges(options: []) { (range, stop) in
```
This could be `range, _`.

xwu commented on Foundation/NSXMLNode.swift:
```diff
@@ -617,8 +617,8 @@ open class XMLNode: NSObject, NSCopying {
                 let possibleParentNodes = grandparent.filter { $0.name == self.parent?.name }
-                let count = possibleParentNodes.reduce(0) {
-                    return $0.0 + 1
+                let count = possibleParentNodes.reduce(0) { (x, y) in
```
Here too, this could be `x, _`.

igor-makarov commented:
It seems like most of those are derived from problems in they way functions in the Swift Standard Library are defined, for example `public func forEach(_ body: ((key: Key, value: Value)) throws -> Void) rethrows` expects a function that receives a tuple, whereas if it was written as `func forEach(_ body: (_ key: Key, _ value: Value) throws -> Void)`, it would expect a function that receives two params.

spevans commented:
Rebased. I added an extra commit with some tidy ups of the closure arguments although of course these are subject to personal taste more than anything.

@parkera Is this ok for merging now? I noticed most of the SE-110 discussion seems to have ended and any possible changes to the parser will be post swift4 now.

parkera commented:
I don't think that's actually the case, or at least I strongly hope it's not, so I still want to hold off on this.

spevans commented:
Closing for now, this is not needed as the SE-110 change is going to be backed out. See https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20170619/037616.html

If any individual closures still generate warnings after the change a new PR can be opened.
